packages/core/src/agent/MessageReceiver.ts

```diff
@@ -63,18 +65,16 @@ export class MessageReceiver {
     const senderKey = unpackedMessage.senderVerkey
     const recipientKey = unpackedMessage.recipientVerkey
 
-    let connection = undefined
+    let connection: ConnectionRecord | null = null
+
+    // Only fetch connection if recipientKey and senderKey are present (AuthCrypt)
     if (senderKey && recipientKey) {
-      // TODO: only attach if theirKey is present. Otherwise a connection that may not be complete, validated or correct will
-      // be attached to the message context. See #76
-      connection = (await this.connectionService.findByVerkey(recipientKey)) || undefined
+      connection = await this.connectionService.findByVerkey(recipientKey)
 
-      // We check whether the sender key is the same as the key we have stored in the connection
-      // otherwise everyone could send messages to our key and we would just accept
-      // it as if it was send by the key of the connection.
+      // Throw error if the recipient key (ourKey) does not match the key of the connection record
       if (connection && connection.theirKey != null && connection.theirKey != senderKey) {
```

Comment:
It doesn't completely suffice, as in this case if `connection.theirKey` is null, it will not throw the error. If theirKey is null, and senderKey isn't (it will never be null due to check above) it will throw the error.

if `connection.theirKey` is null it means the connection does not have their key yet (we've sent an invitation, but didn't receive a request yet). This check is mostly to check if a connection that already has theirKey matches the key of the message we received

Reply:
Agree on the `!==` however